blockchain/contracts/WavGame.sol

```diff
@@ -0,0 +1,270 @@
+
+
+contract WavGame is IWavGame, Ownable, ERC2771Recipient, ReentrancyGuard, ERC165 {
+    using EnumerableSet for EnumerableSet.UintSet;
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    uint256 public entryFee; // In wei
+    uint256 constant internal ENTRY_LEVEL = 1; 
+    IWavContract public wavContract; 
+    address payable public revenueContract; // Platform revenue spliting contract
+
+    mapping(address => IWavGame.Level[]) internal wavGames; // Levels are 0-indexed. i.e level 1 = 0 in level
+    mapping(address => mapping(uint256 => EnumerableSet.AddressSet)) collectors;// Collectors per level per owner game
+    mapping(address => mapping(uint256 => EnumerableSet.UintSet)) burnableSet;// Burnable IDs per level per owner game
+    mapping(address => mapping(uint256 => EnumerableSet.UintSet)) mintableSet;// Mintable IDs per level per owner game
+
+    
+    event Collected(address caller, address to, uint256 amount, uint256 refunded, uint256 totalMinted);
+    event LeveledUp(address caller, address collector, uint256 nextLevel, uint256 totalMinted);
+    event SpecialMint(address caller, address collector, uint256 id, uint256 amount);
+    event PaymentForwarded(address to, uint256 amount);
+    event GameSet(address owner, uint256 level);
+    event LevelSet(address owner, uint256 level);
+
+	// Create all initial artists
+	// create and map the 6 artist level / island to each artist, and define the Levels specific details per artist
+    constructor(IWavContract _wavContract, address _trustedForwarder,uint256 _entryFee, address payable _revenueContract) {
+        wavContract = _wavContract;
+        entryFee = _entryFee;
+        revenueContract = _revenueContract;
+        _setTrustedForwarder(_trustedForwarder);
+    }
+
+    // constructor(IWavContract _wavContract, uint256 _entryFee, address payable _revenueContract) {
+    //     wavContract = _wavContract;
+    //     entryFee = _entryFee;
+    //     revenueContract = _revenueContract;
+    // }
+
+
+    /// @notice This function mints level 1 NFTs to the recipient
+    /// @param _recipient The recipient of the minted NFT
+    /// @param _input ids and their corresponding quantities to mint
+    /// @dev This function does not support gasless, has to be called directly by user, since it requires payment in ETH and refund is possible
+    function collect(address _gameOwner, address _recipient, IDParam[] calldata _input) external override payable nonReentrant {
+        require(msg.value > 0, "PAYMENT_REQUIRED");
+        require(Helper.getLevelIndex(ENTRY_LEVEL) < wavGames[_gameOwner].length, "LEVEL_DOES_NOT_EXIST");
+
+        uint totalPayable = 0;
+        uint totalMinted = 0;
+        uint[] memory mintableIds = new uint[](_input.length); 
+        uint[] memory mintableAmountPerId = new uint[](_input.length);
+
+        // Create mintable IDs and valid quantity based on mintable set for base and msg.value
+        for (uint i; i < _input.length;) { 
```

Comment:
You can store `_input.length` in a local variable to avoid accessing calldata on every iteration.

Reply:
I think it will be cheaper getting from calldata than storing length in memory since this will not be updated during the function execution. 

See [here](https://www.alchemy.com/overviews/solidity-gas-optimization#:~:text=can%20be%20removed.-,8.%20Store%20Data%20in%20calldata%20Instead%20of%20Memory%20for%20Certain%20Function,needs%20to%20be%20updated%20when%20calling%20a%20function%2C%20use%20memory%20instead.,-9.%20Use%20immutable)